src/freadR.c

```diff
@@ -335,7 +335,7 @@ bool userOverride(int8_t *type, lenOff *colNames, const char *anchor, const int
               type[i]=CT_STRING; // e.g. CT_ISO8601_DATE changed to character here so that as.POSIXct treats the date-only as local time in tests 1743.122 and 2150.11
               SET_STRING_ELT(colClassesAs, i, tt);
             }
-          } else {
+          } else if (type[i] != CT_ISO8601_DATE || tt != char_Date) {
             type[i] = typeEnum[w-1];                           // freadMain checks bump up only not down
             if (w==NUT) SET_STRING_ELT(colClassesAs, i, tt);
           }
```

Comment:
What is the consequence of a dangling else here? (As in, if the new condition evaluates to false, it would previously be evaluated.) I think a comment for this line would be useful here.

Reply:
Thanks for the call-out, this is something I wasn't understanding well myself.

The key is we just skip updating `type[i]` for that case. We might consider rewriting this then as:

```c
type[i] = (type[i] == CT_ISO8601_DATE && tt == char_Date) ? type[i] : typeEnum[w-1];
```

The "trouble" comes in the next line, because `w==NUT` in the relevant case here (i.e., `"Date"` is not a known class for C-side coercion).

So with the simple fix, we're back to R-side `as.Date()` coercion.

As noted, we _might_ prefer that (retaining fully back-compatible `Date` output columns), but it'll require a corresponding change in the `colClasses=list()` branch.

So it's down again to whether the result of this PR should be "cols parsed as IDate and requested as Date are returned as IDate, fully avoiding any coercion" or "we now skip the middle step in IDate->char->Date coercion; for maximum efficiency, request IDate instead of Date".